storage/remote/queue_manager.go

```diff
@@ -514,30 +615,48 @@ func (s *shards) sendSamples(samples model.Samples) {
 }
 
 // sendSamples to the remote storage with backoff for recoverable errors.
-func (s *shards) sendSamplesWithBackoff(samples model.Samples) {
+func (s *shards) sendSamplesWithBackoff(samples []tsdb.RefSample) error {
 	backoff := s.qm.cfg.MinBackoff
-	req := ToWriteRequest(samples)
+	req, err := s.qm.buildWriteRequest(samples, s.qm.series)
+	// Failing to build the write request is non-recoverable, since it will
+	// only error if marshaling the proto to bytes fails.
+	if err != nil {
+		return err
+	}
+	for {
+		select {
+		case <-s.ctx.Done():
+			return s.ctx.Err()
+		default:
+			begin := time.Now()
+			err := s.qm.client.Store(s.ctx, req)
+
+			sentBatchDuration.WithLabelValues(s.qm.queueName).Observe(time.Since(begin).Seconds())
+			if err != nil {
+				failedSamplesTotal.WithLabelValues(s.qm.queueName).Add(float64(len(samples)))
+			}
+
+			if err == nil {
+				succeededSamplesTotal.WithLabelValues(s.qm.queueName).Add(float64(len(samples)))
+				now := time.Now()
+				s.qm.sentTimestampMetric.Set(float64(now.UnixNano()) / 1e9)
```

Comment:
Shouldn't this be the time of the sample?  So we can compare the last scraped timestamp to the last sent timestamp, and alert if there is a big backlog,

Reply:
That wouldn't work, as we can take samples back in time. If we want to track that the number of pending bytes would be better.

Reply:
this metric is specifically the time at which we last successfully sent samples, there's another metric for the highest timestamp from a sample that we have successfully sent